src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationWriter.java

```diff
@@ -0,0 +1,203 @@
+    this.mutations = new ArrayList<>();
+  }
+
+  /** @return number of rows copied into Spanner */
+  public int getRowCount() {
+    return this.rowCount;
+  }
+
+  /** Build mutation to add to mutations list with data contained within a CopyData payload */
+  public void buildMutation(ConnectionHandler connectionHandler, byte[] payload) throws Exception {
+    List<CSVRecord> records = parsePayloadData(payload);
+    if (!records.isEmpty()
+        && !payloadFitsInCurrentBatch(records.size() * records.get(0).size(), payload.length)) {
+      writeToSpanner(connectionHandler); // Close out the current batch and reset for the next batch
+    }
+    for (CSVRecord record : records) {
+      // Check that the number of columns in a record matches the number of columns in the table
+      if (record.size() != this.tableColumns.keySet().size()) {
+        writeToSpanner(connectionHandler);
+        throw new SQLException(
+            "Invalid COPY data: Row length mismatched. Expected "
+                + this.tableColumns.keySet().size()
+                + " columns, but only found "
+                + record.size());
+      }
+      WriteBuilder builder = Mutation.newInsertBuilder(this.tableName);
+      // Iterate through all table column to copy into
+      for (String columnName : this.tableColumns.keySet()) {
+        TypeCode columnType = this.tableColumns.get(columnName);
+        String recordValue = "";
+        try {
+          recordValue = record.get(columnName).trim();
+          switch (columnType) {
+            case STRING:
+              builder.set(columnName).to(recordValue);
+              break;
+            case BOOL:
+              builder.set(columnName).to(Boolean.parseBoolean(recordValue));
+              break;
+            case INT64:
+              builder.set(columnName).to(Long.parseLong(recordValue));
+              break;
+            case FLOAT64:
+              builder.set(columnName).to(Double.parseDouble(recordValue));
+              break;
+            case BYTES:
+              builder.set(columnName).to(Byte.parseByte(recordValue));
+              break;
+            case TIMESTAMP:
+              builder.set(columnName).to(com.google.cloud.Timestamp.parseTimestamp(recordValue));
+              break;
+          }
+        } catch (NumberFormatException | DateTimeParseException e) {
+          writeToSpanner(connectionHandler);
+          createErrorFile(payload);
+          throw new SQLException(
+              "Invalid input syntax for type "
+                  + columnType.toString()
+                  + ":"
+                  + "\""
+                  + recordValue
+                  + "\"");
+        } catch (IllegalArgumentException e) {
+          writeToSpanner(connectionHandler);
+          createErrorFile(payload);
+          throw new SQLException("Invalid input syntax for column \"" + columnName + "\"");
+        } catch (Exception e) {
+          writeToSpanner(connectionHandler);
```

Comment:
Same as above: Do we need this if we are currently not supporting copies that exceed the mutation limit?